### DISCUSSION TITLE
Patches with OpenFOAMReader

### DISCUSSION THREAD

**Discussion by GoldenWitch34** [pyvista/pyvista#2059] [Q&A]
Hello,

As part of a school project, I'm working on creating a post-processing workflow for CFD data, particularly for data obtained from OpenFOAM.

As such, I'm using OpenFOAMReader. I tried plotting the velocity field at one of the walls of my domain by loading the corresponding patch, but I get values above zero similar to those I get for the internal mesh at the same area. I find this really surprising since I have a no-slip condition on my wall.

How do the boundaries work exactly with OpenFOAMReader? There might be something I'm not getting here.

Thanks in advance!

**Comment by MatthewFlamm** [comment]
Also take a look at this discussion: https://github.com/pyvista/pyvista/discussions/1869.  Are you using pyvista version 0.33 that has the native support for reading in patch data?

My understanding is that it can depend on how the BC is set in OpenFOAM. For example, if a fixedValue is used, then vtkOpenFOAMReader is able to understand this means exactly a velocity of [0, 0, 0].  No Slip may not be interpreted correctly for some reason.  This has been noted before http://bluecfd.github.io/Core/FAQ/no-slip-surfaces-show-non-zero-velocities-in-paraview/.  Other problems may occur with boundary conditions when using turbulence models.  

I suspect, and this is mostly a hunch, that if the Reader cannot deduce the velocity from the value given in the `U` file, that it extrapolates from the internalMesh.  If you can, check the raw data in the `U` file.

**Reply by MatthewFlamm** [reply]
Also what vtk version are you using?  Can you share `pv.Report()`?  There were _significant_ changes to vtkOpenFOAMReader in 9.1.0.  I think one of the changes is that prior versions _always_ used the internalMesh instead of the actual values of the boundary in some cases.

**Reply by MatthewFlamm** [reply]
See https://gitlab.kitware.com/vtk/vtk/-/issues/18125  and https://gitlab.kitware.com/vtk/vtk/-/issues/16873 which I think were fixed either sometimes in the 9.0.x cycle or for 9.1 (I think the latter).

**Reply by GoldenWitch34** [reply]
Thank you for your reply!

I am using the version of PyVista you mention in that discussion, so I do have access to the patch data directly. That's what I'm using to check the data at the wall directly.

I'm using pyvista.OpenFOAMReader and not vtkOpenFOAMReader, are there any fundamental differences between the two?

The boundary conditions are explicitly set to be noSlip BCs at the walls, this might be the issue. 

Thank you for the link you provided, I didn't know this was an issue with ParaView as well. That inspired me to check my results with ParaView and they do seem to match what I get with my code. I'm using ParaView 5.6.0 by the way.

You might be onto something with that last sentence. In order to plot the streamlines of my flow, I decided to calculate the magnitude of the flow velocity through a postProcess function. The value of the velocity magnitude is of a calculated type with a uniform value of 0 at the wall.

Here's what I get when plotting the velocity magnitude that's provided by ParaView :
<img width="564" alt="Capture_U" src="https://user-images.githubusercontent.com/91612538/150612478-a6de5160-78bd-438a-a99b-5d3fe057127c.PNG">

And here's what I get when plotting the velocity magnitude I calculated :
<img width="562" alt="Capture_mag(U)" src="https://user-images.githubusercontent.com/91612538/150612598-64f99551-0f0e-429c-80e8-874b51ea0b8f.PNG">

Do you know if this problem with ParaView has been fixed?

**Reply by MatthewFlamm** [reply]
The OpenFOAMReader in pyvista wraps the vtkOpenFOAMReader in vtk in a more pythonic interface. All of the logic of reading files and data translation comes from vtk.

Paraview also uses vtk primarily for the underlying functionality.

I don't think you answered which vtk version you are using and if you have tried vtk==9.1.0. The Paraview version you are using would not be using this vtk version either.

**Reply by GoldenWitch34** [reply]
Here's my pv.Report() :
<img width="355" alt="capture" src="https://user-images.githubusercontent.com/91612538/150635745-a49522a2-c277-43a3-9af2-7739e12f5270.PNG">

I haven't tried vtk==9.1.0 yet.

Thanks again for the insight!

**Reply by GoldenWitch34** [reply]
Okay, so I tried downloading vtk 9.1.0, but I'm not sure I understand how to use it. The version of vtk I have with Python is still vtk 9.0.3 and I'm not sure how I can update exactly.

**Reply by MatthewFlamm** [reply]
You need to install it into the python environment.

If you are using pip you can do:
```
pip install vtk==9.1.0
```

If you don't have a reason to be use a dev version of pyvista, Id also recommend doing

```
pip install pyvista==0.33.0
```

**Reply by GoldenWitch34** [reply]
It seems to be working now, thank you!

**Reply by MatthewFlamm** [reply]
@GoldenWitch34 I assume this means that you are seeing the correct values on the surface patch when using vtk==9.1.0 even with the "NoSlip" BC?  If so, I think we should consider making an explicit `Note` or `Warning` about this in the documentation.

**Reply by GoldenWitch34** [reply]
That's right.
